### ISSUE TITLE
pymssql not working with Sybase SQL Anywhere (solution)

### ISSUE THREAD

**Issue by LEALCorentin** [pymssql/pymssql#339]
Hello,

This issue is half an issue half a suggestion.

I founded unclear in the documentation is pymssql supposed to work for Sybase SQL Anywhere or not. As I needed to write a Python script and needed to connect to a Sybase database I had to find a library for it. The Sybase library for Python is no more maintained thus I used pymssql. After a long time trying it without any result I started analysing the code. I realised that, in fact, pymssql is not made for Sybase SQL Anywhere. And this can quite easily be changed.
To help you correct things properly (I haven't time to make it myself, sorry) I will explain you the problems and what to change. First to connect a Sybase SQL Anywhere database, it is needed to use FreeTDS version 5.0 and pymssql do not support it. Secondly, in Sybase the conn_properties parameter is not needed thus I needed to pass to the connection function an empty parameter to avoid the database to reject my connection. This worked perfectly for me using Debian but do not work with ArchLinux I can't find why...

I join you a diff file I created to show you what are the main changes I made. I think it can be made a cleaner manner but you have the big idea.

```
diff -rBu pymssql-2.1.1Old/_mssql.pyx pymssql-2.1.1New/_mssql.pyx
--- pymssql-2.1.1Old/_mssql.pyx	2014-11-25 02:55:04.000000000 +0100
+++ pymssql-2.1.1New/_mssql.pyx	2015-02-26 14:57:22.000000000 +0100
@@ -34,6 +34,7 @@
 # Provide constants missing in FreeTDS 0.82 so that we can build against it
 DEF DBVERSION_71 = 5
 DEF DBVERSION_72 = 6
+DEF DBVERSION_50 = 7
 
 ROW_FORMAT_TUPLE = 1
 ROW_FORMAT_DICT = 2
@@ -515,7 +516,7 @@
                 return 8.0  # Actually 7.1, return 8.0 to keep backward compatibility
             elif version == 8:
                 return 7.0
-            elif version == 6:
+            elif version == 7:
                 return 5.0
             elif version == 4:
                 return 4.2
@@ -1675,6 +1676,8 @@
     """
     if verstr == u'4.2':
         return DBVERSION_42
+    if verstr == u'5.0':
+        return DBVERSION_50
     if verstr == u'7.0':
         return DBVERSION_70
     if verstr == u'7.1':
Seulement dans pymssql-2.1.1New: _mssql.pyx~
diff -rBu pymssql-2.1.1Old/pymssql.pyx pymssql-2.1.1New/pymssql.pyx
--- pymssql-2.1.1Old/pymssql.pyx	2014-11-25 02:55:04.000000000 +0100
+++ pymssql-2.1.1New/pymssql.pyx	2015-02-26 15:00:56.000000000 +0100
@@ -577,7 +577,7 @@
 
 def connect(server='.', user='', password='', database='', timeout=0,
         login_timeout=60, charset='UTF-8', as_dict=False,
-        host='', appname=None, port='1433', conn_properties=None, autocommit=False):
+        host='', appname=None, port='1433', tds_version='5.0', conn_properties=None, autocommit=False):
     """
     Constructor for creating a connection to the database. Returns a
     Connection object.
@@ -629,7 +629,7 @@
     try:
         conn = _mssql.connect(server=server, user=user, password=password,
                               charset=charset, database=database,
-                              appname=appname, port=port,
+                              appname=appname, port=port, tds_version=tds_version,
                               conn_properties=conn_properties)
 
     except _mssql.MSSQLDatabaseException, e:
Seulement dans pymssql-2.1.1New: pymssql.pyx~
diff -rBu pymssql-2.1.1Old/setup.py pymssql-2.1.1New/setup.py
--- pymssql-2.1.1Old/setup.py	2014-11-25 02:55:04.000000000 +0100
+++ pymssql-2.1.1New/setup.py	2015-04-02 17:39:38.000000000 +0200
@@ -147,7 +147,8 @@
 compiler = ccompiler.new_compiler()
 
 _extra_compile_args = [
-    '-DMSDBLIB'
+    '-DMSDBLIB',
+    '-std=c99'
 ]
 
 WINDOWS = False
@@ -208,7 +209,7 @@
             if compiler.has_function('clock_gettime', libraries=['rt']):
                 libraries.append('rt')
 
-usr_local = '/usr/local'
+usr_local = '/usr'
 if osp.exists(usr_local):
     add_dir_if_exists(
         include_dirs,

```

This diff is applicated to the last release (2.1.1)

If you find the reason why it doesn't work on ArchLinux, please tell me I would be very grateful.

**Comment by ramiro** [comment]
From the pymssql.org main page

"
pymssql

A simple database interface for Python that builds on top of FreeTDS to provide a Python DB-API (PEP-249) interface to Microsoft SQL Server.
"

Which is also the first paragraph in the documentation.

My instinct is to close this.

Unfortunately we, the pymssql maintainersm are few and time-constrained and don't have the resources to test against Sybase SQL Anywhere and IMHO it wouldn't  be prudent to announce compatibility with a system we can't test against.

Opinions welcome.

**Comment by rsyring** [comment]
I hate to close it when it's just a small change to which TDS versions we permit.  But, I agree that there just aren't enough maintainers currently for this non-mssql issue to be a priority.

Someday...maybe.

**Comment by rsyring** [comment]
Although...I'm now realizing that to properly test this, we'd need to have way to test against a sybase server.  :(

**Comment by LEALCorentin** [comment]
There is currently no Sybase connection library in Python, it could be an opportunity.

Concerning the testing I'm sorry but I won't be able to help as the database I worked on is professional and not accessible from the outside... The only thing I can do is help you if you have questions on my modifications and assure you that these modification did work perfectly for me.

**Comment by rsyring** [comment]
@LEALCorentin are there free versions of sybase that a developer could install locally?  If so, can you point us to them?

**Comment by LEALCorentin** [comment]
@rsyring http://www.sypron.nl/get_ase_soft.html
This may be what you're looking for. Please note that I didn't test it.

**Comment by ramiro** [comment]
If no volunteer steps up to offer taking care of adding this feature in one month from now, I will close this issue.

**Comment by LEALCorentin** [comment]
If you want I can make the changes I have put in the diff file but I am not sure how to make it in a clean way.
The most problematic thing is about the conn properties. It is a parameter only applicable to MSSQL. To connect a Sybase database it must be set to '' not None.

**Comment by rsyring** [comment]
@LEALCorentin regarding the conn properties, can't you you just pass in `conn_properties=''` if you need it to be an empty string?

**Comment by LEALCorentin** [comment]
@rsyring I can of course, this is how I use it for the moment. But as the library is made for MSSQL I don't really know how MSSQL will handle it if an empty string is given...
I never used pymssql to connect a MSSQL database and I haven't any to try with it.
I think it is important to focus the library on the MSSQL everytime and not Sybase as it is the pourpose of this library.

**Comment by rsyring** [comment]
My point is, the only real patch that I see that needs to be made is support for the 5.0 TDS version?  As long as you can pass in an empty string when needed, then the library doesn't need to default to that.  Just let the user override if needed.

**Comment by ramiro** [comment]
> If no volunteer steps up to offer taking care of adding this feature in one month from now, I will close this issue.

To clarify. What I mean with "take care" is to  any combination of
* setup
* maintain
* fund
* sponsor
* take responsability of maintenance from now on

of a envirnmoent which alloows the development, and CI of pymssql against Sybase Anywhere.

**Comment by rsyring** [comment]
@ramiro I agree with your last post in principle.  However, I wonder if all of that would absolutely be needed.  Given the very small amount of code that it seems like we need to change to support it, I wonder if we could just document support for sybase as "best effort."  IMO, that would mean we would just need:

* a patch to get it working
* the OP to test it for us
* confirmation it doesn't break any of our current tests
* a note in the docs

Either way, I think it's worth keeping this issue open and letting whoever finds it and may eventually care know that there might be options here.

**Comment by knedlsepp** [comment]
Just failed to connect to a TDS 5.0 server using `pytds`, which does not have the FreeTDS dependency, as it turned out it doesn't support the old 5.0 version.
I was hoping that at least a library built on top of FreeTDS could pull it off, but now that I read that `pymssql` doesn't support it either that's quite disappointing.
